transports/janus_websockets.c

```diff
@@ -0,0 +1,906 @@
+}
+
+int janus_websockets_get_api_compatibility(void) {
+	/* Important! This is what your plugin MUST always return: don't lie here or bad things will happen */
+	return JANUS_TRANSPORT_API_VERSION;
+}
+
+int janus_websockets_get_version(void) {
+	return JANUS_WEBSOCKETS_VERSION;
+}
+
+const char *janus_websockets_get_version_string(void) {
+	return JANUS_WEBSOCKETS_VERSION_STRING;
+}
+
+const char *janus_websockets_get_description(void) {
+	return JANUS_WEBSOCKETS_DESCRIPTION;
+}
+
+const char *janus_websockets_get_name(void) {
+	return JANUS_WEBSOCKETS_NAME;
+}
+
+const char *janus_websockets_get_author(void) {
+	return JANUS_WEBSOCKETS_AUTHOR;
+}
+
+const char *janus_websockets_get_package(void) {
+	return JANUS_WEBSOCKETS_PACKAGE;
+}
+
+int janus_websockets_send_message(void *transport, void *request_id, gboolean admin, json_t *message) {
+	if(message == NULL)
+		return -1;
+	if(transport == NULL) {
+		g_free(message);
+		return -1;
+	}
+	janus_websockets_client *client = (janus_websockets_client *)transport;
+	if(!client->context || !client->wsi) {
+		g_free(message);
+		return -1;
+	}
+	/* Make sure this is not related to a closed WebSocket session */
+	janus_mutex_lock(&old_wss_mutex);
+	if(g_list_find(old_wss, client) == NULL) {
+		janus_mutex_lock(&client->mutex);
+		/* Convert to string and enqueue */
+		char *payload = json_dumps(message, JSON_INDENT(3) | JSON_PRESERVE_ORDER);
+		g_async_queue_push(client->messages, payload);
+		libwebsocket_callback_on_writable(client->context, client->wsi);
+		janus_mutex_unlock(&client->mutex);
+	}
+	janus_mutex_unlock(&old_wss_mutex);
+	json_decref(message);
+	return 0;
+}
+
+void janus_websockets_session_created(void *transport, guint64 session_id) {
+	/* We don't care */
+}
+
+void janus_websockets_session_over(void *transport, guint64 session_id, gboolean timeout) {
+	if(transport == NULL || !timeout)
+		return;
+	/* We only care if it's a timeout: if so, close the connection */
+	janus_websockets_client *client = (janus_websockets_client *)transport;
+	if(!client->context || !client->wsi)
+		return;
+	/* Make sure this is not related to a closed WebSocket session */
+	janus_mutex_lock(&old_wss_mutex);
+	if(g_list_find(old_wss, client) == NULL) {
+		janus_mutex_lock(&client->mutex);
+		client->session_timeout = 1;
+		libwebsocket_callback_on_writable(client->context, client->wsi);
+		janus_mutex_unlock(&client->mutex);
+	}
+	janus_mutex_unlock(&old_wss_mutex);
+}
+
+
+/* Thread */
+void *janus_websockets_thread(void *data) {
+	JANUS_LOG(LOG_INFO, "WebSocket thread started\n");
+	while(g_atomic_int_get(&initialized) && !g_atomic_int_get(&stopping)) {
+		/* libwebsockets is single thread, we cycle through events here */
+		if(wss != NULL)
+			libwebsocket_service(wss, 100);
+		if(swss != NULL)
+			libwebsocket_service(swss, 100);
+		if(admin_wss != NULL)
+			libwebsocket_service(admin_wss, 100);
+		if(admin_swss != NULL)
+			libwebsocket_service(admin_swss, 100);
```

Comment:
It appears that, if someone enables all the websocket endpoints, it could take up to 300 milliseconds for a message to be noticed. Might be a good idea to divide a total cycle time by the number of enabled endpoints, or start a thread for each enabled endpoint.